File: web/src/domain/factory/DiagnosticoFactory.ts
import { Evaluacion } from "../models/Evaluacion";

export interface DiagnosticoEngine {
  generate(evaluacion: Evaluacion): string;
}

class StandardEngine implements DiagnosticoEngine {
  generate(evaluacion: Evaluacion): string {
    return "Diagnóstico Estándar (Web)";
  }
}

class AdvancedEngine implements DiagnosticoEngine {
  generate(evaluacion: Evaluacion): string {
    return "Diagnóstico Avanzado (Web)";
  }
}

export class DiagnosticoFactory {
  static create(role: string): DiagnosticoEngine {
    if (role === "admin" || role === "optometrista") {
      return new AdvancedEngine();
    }
    return new StandardEngine();
  }
}

File: app/src/main/java/com/example/optoapp/domain/factory/DiagnosticoFactory.kt
package com.example.optoapp.domain.factory

import com.example.optoapp.domain.model.EvaluacionModel

interface DiagnosticoEngine {
    fun generateDiagnostico(evaluacion: EvaluacionModel): String
}

class StandardDiagnosticoEngine : DiagnosticoEngine {
    override fun generateDiagnostico(evaluacion: EvaluacionModel): String {
        return "Diagnóstico Estándar basado en refracción"
    }
}

class AdvancedDiagnosticoEngine : DiagnosticoEngine {
    override fun generateDiagnostico(evaluacion: EvaluacionModel): String {
        return "Diagnóstico Avanzado con IA y análisis binocular"
    }
}

class DiagnosticoFactory {
    fun createEngine(role: String): DiagnosticoEngine {
        return when (role) {
            "admin", "optometrista" -> AdvancedDiagnosticoEngine()
            else -> StandardDiagnosticoEngine()
        }
    }
}
